slang.h

```diff
@@ -450,6 +450,178 @@ convention for interface methods.
+
+#define SLANG_MAKE_PRIMARY_STRUCT_TAG(CATEGORY, TYPE_ID, MAJOR, MINOR) slang::StructTag(slang::StructTagInt(slang::kStructTagPrimaryMask) | (slang::StructTagInt(CATEGORY) << 24) | (slang::StructTagInt(TYPE_ID) << 16) | (slang::StructTagInt(MAJOR) << 8) | slang::StructTagInt(MINOR))
+#define SLANG_MAKE_EXTENSION_STRUCT_TAG(CATEGORY, TYPE_ID, MAJOR, MINOR) slang::StructTag((slang::StructTagInt(CATEGORY) << 24) | (slang::StructTagInt(TYPE_ID) << 16) | (slang::StructTagInt(MAJOR) << 8) | slang::StructTagInt(MINOR))
+
+/* 
+`PrimaryStruct`s are structs that are passed directly into API calls. They always contain a 'structType' that identifies the exact version
+and type of that is being passed in. `PrimaryStruct`s can also specify optional `ExtensionStruct` types that modify and/or add to the values
+in the `PrimaryStruct`.
+
+`ExtensionStruct` cannot contain optional other extensions as PrimaryStructs do - all required extensions have to be specified via the list set on the
+`PrimaryStruct`.
+
+`ExtensionStruct` is typically used when it is necessary to provide some special additional information that is not appropriate to place within
+the `PrimaryStruct`. This also provides a mechanism such that a `PrimaryStruct` derived type does not need to include all the fields that will
+every be needed. Cross cutting aspects can have their own uniquely identified structs.
+
+As a mechanism to provide extensibility without having to modify a struct or to allow optional and/or additional information through
+ExtensionStruct set via the `exts` and `extsCount` members.
+
+A type that is ABI compatible with this mechanism must start with exactly the same fields as specified.
+When using extensions, the exts and extsCount should be set. `exts` points to the extension ids (which must be at the start of the types)
+to provide some kind of type safely without requiring inheritance.
+*/
+
+#define SLANG_EXTENSION_TAGGED_STRUCT_IMPL(TYPE_NAME, CATEGORY, TYPE_ID, MAJOR, MINOR) \
+    typedef TYPE_NAME ThisType; \
+    typedef slang::ExtensionTag Tag; \
+    static const slang::StructTag kStructTag = SLANG_MAKE_EXTENSION_STRUCT_TAG(CATEGORY, TYPE_ID, MAJOR, MINOR); \
+    slang::StructTag structTag = kStructTag; \
+    slang::StructSize structSize = slang::StructSize(sizeof(ThisType));
+
+#define SLANG_PRIMARY_TAGGED_STRUCT_IMPL(TYPE_NAME, CATEGORY, TYPE_ID, MAJOR, MINOR) \
+    typedef TYPE_NAME ThisType; \
+    typedef ::slang::PrimaryTag Tag; \
+    static const ::slang::StructTag kStructTag = SLANG_MAKE_PRIMARY_STRUCT_TAG(CATEGORY, TYPE_ID, MAJOR, MINOR); \
+    slang::StructTag structTag = kStructTag; \
+    slang::StructSize structSize = slang::StructSize(sizeof(ThisType)); \
+    const slang::StructTag** exts = nullptr; \
+    int32_t extsCount = 0; 
+
+#define SLANG_API_TYPE_ENUM(x) x,
+
+struct TaggedStructBase
+{
+    StructTag structTag;            ///< Identity for the type
+    StructSize structSize;             ///< Size of the type in bytes
```

Comment:
Why are we *also* including `structSize` if we already have `structTag`? I very much thought of this as an either/or.

My thought had been that the validation logic traffics in two main cases:

1. The `structTag` is one that is allowed + recognized (and thus have a declaration for) and we can freely use that declaration directly to copy things: `ConcreteType& concrete = *(ConcreteType*)structPtr;`

2. The `structTag` is one that is not allowed (whether recognized or not), so we simply diagnose an error about the mismatch, which relies on name/version/whatever information from the tag (and doesn't care about the actual data in the struct)

I can't see how either of those cases is improved by having a `structSize` field, and having both seems wasteful to me. I'm happy to be convinced otherwise, of course.

Reply:
Having the structSize was added so I could support arrays of structs as opposed to arrays of pointers to structs. Having it allows for struct from the 'future' or the 'past' relative to current.